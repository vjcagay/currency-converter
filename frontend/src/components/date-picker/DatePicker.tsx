import AirDatepicker from "air-datepicker";
import "air-datepicker/air-datepicker.css";
import localeEn from "air-datepicker/locale/en";
import React, { useEffect, useRef, useState } from "react";
import styles from "./styles.module.css";
import CalendarIcon from "./assets/calendar.svg";
import ResetIcon from "./assets/reset.svg";

type Props = {
  className?: string;
  min?: string;
  max?: string;
  from: string;
  to: string;
  onChange: (range: { from: string; to: string }) => void;
};

type CalendarProps = {
  min?: string;
  max?: string;
  from: string;
  to: string;
  onChange: (range: { from: string; to: string }) => void;
  onCancel: () => void;
};

const formatDateString = (date: Date) => {
  // Month starts at zero so we +1
  return `${date.getFullYear()}-${(date.getMonth() + 1).toString(10).padStart(2, "0")}-${date
    .getDate()
    .toString(10)
    .padStart(2, "0")}`;
};

const Calendar = (props: CalendarProps) => {
  const divRef = useRef<HTMLDivElement>();
  const calendarRef = useRef<AirDatepicker<HTMLDivElement>>();

  useEffect(() => {
    if (divRef.current) {
      calendarRef.current = new AirDatepicker(divRef.current, {
        classes: styles.calendar,
        inline: true,
        locale: localeEn,
        navTitles: {
          days: "MMM yyyy", // Day view title will show up as something like `Jan 2023`
        },
        range: true,
        buttons: [
          {
            content: "Cancel",
            onClick: props.onCancel,
          },
          {
            content: "Confirm",
            onClick(picker) {
              if (picker.selectedDates.length === 2) {
                props.onChange({
                  from: formatDateString(picker.selectedDates[0]),
                  to: formatDateString(picker.selectedDates[1]),
                });
              }
            },
          },
        ],
        onSelect(picker) {
          const confirmButton = picker.datepicker.$el.getElementsByClassName("air-datepicker-button").item(1);

          // Disable the `Confirm` button when only one date is selected
          if (Array.isArray(picker.date) && picker.date.length !== 2) {
            confirmButton.setAttribute("disabled", "true");
          } else {
            confirmButton.removeAttribute("disabled");
          }
        },
        minDate: props.min,
        maxDate: props.max,
        selectedDates: [props.from, props.to],
      });
    }

    return () => calendarRef.current.destroy();
  }, [divRef]);

  return <div ref={divRef} />;
};

const presetRanges = ["1W", "1M", "6M"];

const DatePicker = (props: Props) => {
  const [selectedPresetRange, setSelectedPresetRange] = useState("");

  useEffect(() => {
    if (selectedPresetRange) {
      const date = new Date();

      let from = "";
      let to = formatDateString(date);

      if (selectedPresetRange === "1W") {
        date.setDate(date.getDate() - 7);
      }

      if (selectedPresetRange === "1M") {
        date.setMonth(date.getMonth() - 1);
      }

      if (selectedPresetRange === "6M") {
        date.setMonth(date.getMonth() - 6);
      }

      from = formatDateString(date);

      props.onChange({ from, to });
    }
  }, [selectedPresetRange]);

  const [isCustomRangePickerOpen, setIsCustomRangePickerOpen] = useState(false);

  return (
    <div className={`${styles.wrapper} ${props.className || ""}`}>
      <fieldset className={`${styles.fieldset} ${styles.presetRangeFieldset}`}>
        {presetRanges.map((range) => (
          <label key={range} className={styles.presetRangeLabel} data-testid={`date-picker-range-${range}`}>
            <input
              type="radio"
              checked={selectedPresetRange === range}
              onChange={(event) => setSelectedPresetRange(range)}
            />
            <span>{range}</span>
          </label>
        ))}
      </fieldset>
      <fieldset className={`${styles.fieldset} ${styles.customRangeFieldset}`}>
        <button
          className={styles.customRangeFieldsetButton}
          data-testid="date-picker-open-calendar"
          onClick={() => setIsCustomRangePickerOpen(true)}
        >
          <img src={CalendarIcon} alt="calendar icon" />
        </button>
        {!selectedPresetRange && (
          <>
            <span
              className={styles.customRangeText}
              data-testid="date-picker-custom-range"
            >{`${props.from} - ${props.to}`}</span>
            <button
              className={`${styles.customRangeFieldsetButton} ${styles.customRangeResetButton}`}
              data-testid="date-picker-reset-range"
              onClick={() => setSelectedPresetRange("1W")}
            >
              <img src={ResetIcon} alt="reset icon" />
            </button>
          </>
        )}
      </fieldset>
      {isCustomRangePickerOpen && (
        <div className={styles.customRangeBackdrop}>
          <Calendar
            min={props.min}
            max={props.max}
            from={props.from}
            to={props.to}
            onCancel={() => setIsCustomRangePickerOpen(false)}
            onChange={(range) => {
              props.onChange(range);
              setIsCustomRangePickerOpen(false);
              setSelectedPresetRange("");
            }}
          />
        </div>
      )}
    </div>
  );
};

export default DatePicker;
