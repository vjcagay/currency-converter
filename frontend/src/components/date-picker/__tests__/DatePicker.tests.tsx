import { fireEvent, render } from "@testing-library/react";
import React from "react";
import DatePicker from "../DatePicker";

// When selecting the preset ranges, DatePicker selects the current date as the `to` value
// Since we are mocking the Date object, `to` will be 2023 Oct 10.
jest.useFakeTimers();
jest.setSystemTime(new Date("2023-10-10").getTime());

afterAll(() => {
  jest.useRealTimers();
});

describe("<DatePicker />", () => {
  it("should render correctly", () => {
    const { asFragment } = render(<DatePicker from="2022-10-03" to="2022-10-10" onChange={() => {}} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("should return the correct `from` and `to` dates when selecting the preset ranges", async () => {
    const onChangeMock = jest.fn();

    const { findByTestId } = render(<DatePicker from="2023-01-01" to="2023-12-31" onChange={onChangeMock} />);

    fireEvent.click(await findByTestId("date-picker-range-1W"));
    expect(onChangeMock).toHaveBeenCalledWith({ from: "2023-10-03", to: "2023-10-10" });

    fireEvent.click(await findByTestId("date-picker-range-1M"));
    expect(onChangeMock).toHaveBeenCalledWith({ from: "2023-09-10", to: "2023-10-10" });

    fireEvent.click(await findByTestId("date-picker-range-6M"));
    expect(onChangeMock).toHaveBeenCalledWith({ from: "2023-04-10", to: "2023-10-10" });
  });

  it("should return the correct `from` and `to` dates when selecting the ranges using the calendar", async () => {
    const onChangeMock = jest.fn();

    const { findByTestId, findByText } = render(
      <DatePicker from="2023-01-01" to="2023-12-31" onChange={onChangeMock} />,
    );

    fireEvent.click(await findByTestId("date-picker-open-calendar"));
    fireEvent.click((await findByText("Confirm")).parentElement);
    expect(onChangeMock).toHaveBeenCalledWith({ from: "2023-01-01", to: "2023-12-31" });
    expect((await findByTestId("date-picker-custom-range")).textContent).toBe("2023-01-01 - 2023-12-31");
  });

  it("should revert to 1w range when clicking the reset button", async () => {
    const onChangeMock = jest.fn();

    const { findByTestId, findByText } = render(
      <DatePicker from="2023-01-01" to="2023-12-31" onChange={onChangeMock} />,
    );

    fireEvent.click(await findByTestId("date-picker-open-calendar"));
    fireEvent.click((await findByText("Confirm")).parentElement);
    fireEvent.click(await findByTestId("date-picker-reset-range"));
    expect(onChangeMock).toHaveBeenCalledWith({ from: "2023-10-03", to: "2023-10-10" });
  });
});
