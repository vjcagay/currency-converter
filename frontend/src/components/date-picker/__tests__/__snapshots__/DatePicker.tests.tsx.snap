// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatePicker /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="wrapper "
  >
    <fieldset
      class="fieldset presetRangeFieldset"
    >
      <label
        class="presetRangeLabel"
        data-testid="date-picker-range-1W"
      >
        <input
          type="radio"
        />
        <span>
          1W
        </span>
      </label>
      <label
        class="presetRangeLabel"
        data-testid="date-picker-range-1M"
      >
        <input
          type="radio"
        />
        <span>
          1M
        </span>
      </label>
      <label
        class="presetRangeLabel"
        data-testid="date-picker-range-6M"
      >
        <input
          type="radio"
        />
        <span>
          6M
        </span>
      </label>
    </fieldset>
    <fieldset
      class="fieldset customRangeFieldset"
    >
      <button
        class="customRangeFieldsetButton"
        data-testid="date-picker-open-calendar"
      >
        <img
          alt="calendar icon"
          src="test-file-stub"
        />
      </button>
      <span
        class="customRangeText"
        data-testid="date-picker-custom-range"
      >
        2022-10-03 - 2022-10-10
      </span>
      <button
        class="customRangeFieldsetButton customRangeResetButton"
        data-testid="date-picker-reset-range"
      >
        <img
          alt="reset icon"
          src="test-file-stub"
        />
      </button>
    </fieldset>
  </div>
</DocumentFragment>
`;
